### ISSUE TITLE
Deep Learning model for sentiment analysis

### ISSUE THREAD

**Issue by MadaraUchiha-314** [intuit/fawkes#15]
Currently Fawkes uses [Vader from NLTK](https://www.nltk.org/_modules/nltk/sentiment/vader.html) to do sentiment analysis.

There are more powerful tools available for sentiment analysis which give better results than Vader.

One such model is:
- BERT: [Link to article](https://towardsdatascience.com/multi-class-sentiment-analysis-using-bert-86657a2af156)

**Comment by rubyruins** [comment]
I would like to work on this. Could you explain the issue a bit more?

**Comment by MadaraUchiha-314** [comment]
Sure @rubyruins.

Fawkes starts from a raw user review: [sample-mint/appstore-raw-feedback.json]

Then it parses it and converts to a standard format: [sample-mint/parsed-user-feedback.json]

Then it runs algorithms on it like:
- Sentiment Analysis:  [sentiment.py]
- Classification:  [text_match.py], [lstm_classifier.py]

[sample-mint/processed-user-feedback.json] is the output after running those algorithms.

Currently [sentiment.py] uses Vader which is model from NLTK, but there are much better models available which are based on deep learning which do a much better job at sentiment analysis.

The aim of this issue is to use those to provide better scores.

[sample-mint/processed-user-feedback.json]: https://github.com/intuit/fawkes/blob/master/data/processed_data/sample-mint/processed-user-feedback.json
[sample-mint/appstore-raw-feedback.json]: https://github.com/intuit/fawkes/blob/master/data/raw_data/sample-mint/appstore-raw-feedback.json
[sample-mint/parsed-user-feedback.json]: https://github.com/intuit/fawkes/blob/master/data/parsed_data/sample-mint/parsed-user-feedback.json
[sentiment.py]: https://github.com/intuit/fawkes/blob/master/src/algorithms/sentiment.py
[text_match.py]: https://github.com/intuit/fawkes/blob/master/src/algorithms/text_match/text_match.py
[lstm_classifier.py]: https://github.com/intuit/fawkes/blob/master/src/algorithms/lstm_classifier/lstm_classifier.py

**Comment by rubyruins** [comment]
Thanks for the clarification! I have worked with VADER and have some basic experience in LSTMs. I'm willing to learn and build a BERT model for this. May I work on this for Hacktoberfest?

**Comment by MadaraUchiha-314** [comment]
@rubyruins Sure thing. I will assign it to you.

**Comment by rubyruins** [comment]
Thank you! I will start working on it soon and reach out in case of doubts, hope that's okay.

**Comment by rubyruins** [comment]
Hi, I have started working on this and will send a PR soon.

**Comment by rubyruins** [comment]
@MadaraUchiha-314, could you please add the `Hacktoberfest` label to the repository? Since the rules were changed,  contributions won't be counted for Hacktoberfest without the label!

**Comment by rubyruins** [comment]
@MadaraUchiha-314, quick question - where do I get the fawkes reviews dataset to train the model on? The one in the article uses pretrained weights and then trains it on the Yelp review dataset. 
Also, which version of BERT should I use from these?

![image](https://user-images.githubusercontent.com/50259869/95009121-1b8e1480-063d-11eb-993b-2972930f40e3.png)

Large uncased looks good to me, what do you think?

**Comment by rubyruins** [comment]
Or should I train it on Yelp reviews itself? Would that be relevant?

**Comment by rubyruins** [comment]
It looks like my laptop's drivers are not up to date, hence I'm not able to use tensorflow and get the model working. I think I'll take a raincheck on this issue and unassign myself, hope that's okay. Sorry!

**Comment by MadaraUchiha-314** [comment]
@rubyruins You can use Google Colab. https://colab.research.google.com/

**Comment by spencerderecat** [comment]
I'm taking a look into this currently

**Comment by spencerderecat** [comment]
@MadaraUchiha-314 what are the classes that we want to have here?

**Comment by MadaraUchiha-314** [comment]
@spencerderecat Are you referring to sentiment classes ? Positive, Negative, Neutral should be fine.

**Comment by spencerderecat** [comment]
is there a training set of reviews?

**Comment by spencerderecat** [comment]
@MadaraUchiha-314 we cant train an ml model if we don't have a training set. I can train an ml model based on rating, but there really isn't any point in doing this as we will always have access to the rating. Best we can do without a training set is find the best pertained model (something like Vader) to use.

**Comment by spencerderecat** [comment]
https://colab.research.google.com/drive/19Q65S0c6VHuyIOxCFe4Vk0hmsF4HiAgj?usp=sharing here is the progress I have made. It uses berts and hugging face to train a model on data that has user ratings (4/5 = positive, 0/1/2/3 = negative) - it can then predict a positive or negative review for data w/o user reviews (ie twitter posts). The real work that needs to be done is integrating this into the application. This still needs to be done.
